roles/keycloak/tasks/main.yml

```diff
@@ -0,0 +1,79 @@
+# Copyright (c) 2022 VEXXHOST, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License. You may obtain
+# a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+
+- name: Get the Kuberentes service for Percona XtraDB Cluster
+  run_once: true
+  kubernetes.core.k8s_info:
+    kind: Service
+    name: "{{ openstack_helm_endpoints.oslo_db.hosts.default }}"
+    namespace: openstack
+  register: _pxc_service
+
+- name: Install Kubernetes python package
+  ansible.builtin.pip:
+    name: PyMySQL
+
+- name: Create Keycloak database
+  run_once: true
+  community.mysql.mysql_db:
+    login_host: "{{ _pxc_service.resources[0].spec.clusterIP }}"
+    login_user: root
+    login_password: "{{ openstack_helm_endpoints.oslo_db.auth.admin.password }}"
+    name: "{{ keycloak_database_name }}"
+
+- name: Create a Keycloak user
+  community.mysql.mysql_user:
+    login_host: "{{ _pxc_service.resources[0].spec.clusterIP }}"
+    login_user: root
+    login_password: "{{ openstack_helm_endpoints.oslo_db.auth.admin.password }}"
+    name: "{{ keycloak_database_username }}"
+    password: "{{ keycloak_database_password }}"
+    host: "%"
+    priv: "{{ keycloak_database_name }}.*:ALL"
+
+- name: Disable pxc strict mode
+  community.mysql.mysql_query:
+    login_host: "{{ _pxc_service.resources[0].spec.clusterIP }}"
+    login_user: root
+    login_password: "{{ openstack_helm_endpoints.oslo_db.auth.admin.password }}"
+    query: "set global pxc_strict_mode='PERMISSIVE'"
+
+- name: Deploy Helm chart
+  run_once: true
+  kubernetes.core.helm:
+    name: "{{ keycloak_helm_release_name }}"
+    chart_ref: "{{ keycloak_helm_chart_ref }}"
+    release_namespace: "{{ keycloak_helm_release_namespace }}"
+    create_namespace: true
+    kubeconfig: /etc/kubernetes/admin.conf
+    wait: true
+    values: "{{ _keycloak_helm_values | combine(keycloak_helm_values, recursive=True) }}"
+
+- name: Disable pxc strict mode
+  community.mysql.mysql_query:
+    login_host: "{{ _pxc_service.resources[0].spec.clusterIP }}"
+    login_user: root
+    login_password: "{{ openstack_helm_endpoints.oslo_db.auth.admin.password }}"
+    query: "set global pxc_strict_mode='ENFORCING'"
+
+- name: Wait until Keycloak ready
```

Comment:
yes, there is.
Once we move keycloak health check into `keycloak_client` and `keycloak_realm` role, we cannot rely on the ready status of K8s resource(sts here). We should check if keycloak server is functioning using curl to `keycloak_host` endpoint.
ofc, we can check keycloak health check endpoints https://www.keycloak.org/server/health  in those roles in advance. But health check endpoints are disabled in default so it should be enabled explicitly.

I thought it is more reasonable to confirm keycloak is ready in keycloak role and in those client and realm roles, it should be fail if the provided keycloak endpoint is not responding or unhealthy instead of waiting.
One more thing is keycloak server bootstrapping takes a bit long time during db migration task.

Reply:
@okozachenko1203  but since we have retries in the task, wouldn't the retry just eventually work?